File: interfatecs-maratona-de-programacao/ANTIGAS/07. Ordenação - Conceitos e Algoritmos I/ordenacao.py
'''
╔══════════════════════════════════════════════════════════════════════════════╗
║ Instituição   :  Faculdade de Tecnologia de São Paulo                        ║
║ Departamento  :  Tecnologia da Informação                                    ║
║ Curso         :  Análise e Desenvolvimento de Sistemas                       ║
║ Autor         :  Lucio Nunes de Lira                                         ║
╠══════════════════════════════════════════════════════════════════════════════╣
║ Palestra      :  Ordenação: Conceitos & Algoritmos                           ║
║ Programa      :  Algoritmos de ordenação por comparação                      ║
║ Linguagem     :  Python 3                                                    ║
║ Compilador    :  CPython (3.6.5)                                             ║
║ Versão        :  A (Rev. 1)                                                  ║
╚══════════════════════════════════════════════════════════════════════════════╝
'''

import random
random.seed()

'''───────────────────────────────────────────────
 Funções auxiliares
───────────────────────────────────────────────'''
def aleatorio(i, f):
    return random.randint(i, f)

def crescente(v, n):
    i = 1
    while i < n and v[i-1] <= v[i]: i += 1
    return i == n

def troca(v, i, j):
    v[i],v[j] = v[j],v[i]

'''───────────────────────────────────────────────
 Bubble Sort
───────────────────────────────────────────────'''
def empurra(v, n):
    for i in range(n-1):
        if v[i] > v[i+1]: troca(v, i, i+1)

def bubble_sort(v, n):
    for i in range(n, 1, -1): empurra(v, i)

'''───────────────────────────────────────────────
 Selection Sort
───────────────────────────────────────────────'''
def seleciona(v, n):
    m = 0
    for i in range(1, n):
        if v[i] > v[m]: m = i
    return m

def selection_sort(v, n):
    for i in range(n, 1, -1):
        m = seleciona(v, i)
        troca(v, i-1, m)
        
'''───────────────────────────────────────────────
 Insertion Sort
───────────────────────────────────────────────'''
def insere(x, v, n):
    i = n-2
    while i>=0 and x<v[i]:
        v[i+1] = v[i]
        i -= 1
    v[i+1] = x

def insertion_sort(v, n):
    for i in range(1, n): insere(v[i], v, i+1)

'''───────────────────────────────────────────────
 Heap Sort
───────────────────────────────────────────────'''
def desce(i, v, n):
    while 2*i + 1 < n: # enquanto há filho a esquerda
        j = 2*i + 1    # guarda o índice do filho da esquerda
        if j+1 < n and v[j+1] > v[j]: j += 1 # se há filho da direita e ele é maior que o da esquerda, guarda o índice dele 
        if v[i] >= v[j]: break # se o pai é maior ou igual ao índice do maior filho, não há o que corrigir
        troca(v, i, j) # troca-se o pai e o filho de posição (o pai é rebaixado)
        i = j          # agora corrigiremos o pai a partir de sua nova posição (antigo índice do filho)

def amontoa(v):
    n = len(v)
    for i in reversed(range(0, n//2)): # parte do último pai e vai voltando
        desce(i, v, n) # ajusta o pai na posição correta

def heap_sort(v):
    amontoa(v) # transforma o vetor em um heap
    n = len(v) # guarda o tamanho do heap
    for i in reversed(range(1, n)): # para cada posição do heap (começando da última)
        troca(v, 0, i) # troca o item máximo com o último do vetor
        desce(0, v, i) # corrigi o item do topo

'''───────────────────────────────────────────────
 Tree Sort
───────────────────────────────────────────────'''
def insere_arvore(x, a):
    if a == None: return (None, x, None)
    if x <= a[1]: return (insere_arvore(x, a[0]), a[1], a[2])
    return (a[0], a[1], insere_arvore(x, a[2]))

def lista_para_arvore(v):
    a = None
    for x in v: a = insere_arvore(x, a)
    return a

def arvore_para_lista(a):
    if a == None: return []
    return arvore_para_lista(a[0]) + [a[1]] + arvore_para_lista(a[2])

def tree_sort(v):
    w = arvore_para_lista(lista_para_arvore(v))
    for i in range(len(w)): v[i] = w[i]

'''───────────────────────────────────────────────
 Merge Sort
───────────────────────────────────────────────'''
def intercala(v, i, m, f):
    a = i
    b = m+1
    w = []
    while a <= m and b <= f:
        if v[a] <= v[b]:
            w.append(v[a])
            a += 1
        else:
            w.append(v[b])
            b += 1
    while a <= m:
        w.append(v[a])
        a += 1
    while b <= f:
        w.append(v[b])
        b += 1
    for e in w:
        v[i] = e
        i += 1

def merge_sort(v, i, f):
    if i == f: return
    m = (i + f) // 2
    merge_sort(v, i, m)
    merge_sort(v, m+1, f)
    intercala(v, i, m, f)

'''───────────────────────────────────────────────
 Quick Sort (determinístico/probabilístico)
───────────────────────────────────────────────'''
def particiona(v, i, f):
    #troca(v, i, aleatorio(i, f))
    e = i+1
    d = f
    while e <= d:
        while e <= d and v[e] <= v[i]: e += 1
        while v[d] > v[i]: d -= 1
        if e < d:
            troca(v, e, d)
            e += 1
            d -= 1
    troca(v, i, d)
    return d

def quick_sort(v, i, f):
    if i >= f: return
    p = particiona(v, i, f)
    quick_sort(v, i, p-1)
    quick_sort(v, p+1, f)

'''───────────────────────────────────────────────
 Da própria linguagem
───────────────────────────────────────────────'''
v = [aleatorio(1,2**32) for i in range(10**5)]

# Feito no próprio vetor.
v.sort()
v.sort(reverse=True)
v.sort(key=lambda x: x[0])

# Feito em uma cópia do vetor original.
w = sorted(v)
w = sorted(v, reverse=True)
w = sorted(v, key=lambda x: x[0])

File: interfatecs-maratona-de-programacao/ANTIGAS/07. Ordenação - Conceitos e Algoritmos I/ordenacao.c
/*
╔══════════════════════════════════════════════════════════════════════════════╗
║ Instituição   :  Faculdade de Tecnologia de São Paulo                        ║
║ Departamento  :  Tecnologia da Informação                                    ║
║ Curso         :  Análise e Desenvolvimento de Sistemas                       ║
║ Autor         :  Lucio Nunes de Lira                                         ║
╠══════════════════════════════════════════════════════════════════════════════╣
║ Palestra      :  Ordenação: Conceitos & Algoritmos                           ║
║ Programa      :  Algoritmos de ordenação por comparação                      ║
║ Linguagem     :  C99                                                         ║
║ Compilador    :  Pelles C (8.00.60)                                          ║
║ Versão        :  A (Rev. 0)                                                  ║
╚══════════════════════════════════════════════════════════════════════════════╝
*/

#include <stdio.h>
#include <stdlib.h>

typedef int Tipo;

/*────────────────────────────────────────────────
 Funções auxiliares
────────────────────────────────────────────────*/
int aleatorio(int i, int f) {
    return rand() % (f - i + 1) + i;
}

int crescente(Tipo v[], int n) {
    int i = 1;
    while (i < n && v[i-1] <= v[i]) i++;
    return i == n;
}

void troca(Tipo v[], int i, int j) {
    Tipo x = v[i];
    v[i] = v[j];
    v[j] = x;
}

/*────────────────────────────────────────────────
 Bubble Sort
────────────────────────────────────────────────*/
void empurra(Tipo v[], int n) {
    for (int i = 0; i < n-1; i++)
        if (v[i] > v[i+1]) troca(v, i, i+1);
}

void bubble_sort(Tipo v[], int n) {
    for (int i = n; i > 1; i--) empurra(v, i);
}

/*────────────────────────────────────────────────
 Selection Sort
────────────────────────────────────────────────*/
int seleciona(Tipo v[], int n) {
    int m = 0;
    for (int i = 1; i < n; i++)
        if (v[i] > v[m]) m = i;
    return m;
}

void selection_sort(Tipo v[], int n) {
    for (int i = n; i > 1; i--) {
        int m = seleciona(v, i);
        troca(v, i-1, m);
    }
}

/*────────────────────────────────────────────────
 Insertion Sort
────────────────────────────────────────────────*/
void insere(Tipo x, Tipo v[], int n) {
    int i = n-2;
    while (i >= 0 && x < v[i]) {
        v[i+1] = v[i];
        i--;
    }
    v[i+1] = x;
}

void insertion_sort(Tipo v[], int n) {
    for (int i = 1; i < n; i++) insere(v[i], v, i+1);
}

/*────────────────────────────────────────────────
 Merge Sort
────────────────────────────────────────────────*/
void intercala(Tipo v[], int i, int m, int f) {
    int a = i, b = m+1, c = 0;
    Tipo *w = malloc((f-i+1) * sizeof(Tipo));
    while (a <= m && b <= f) w[c++] = v[a] <= v[b] ? v[a++] : v[b++];
    while (a <= m) w[c++] = v[a++];
    while (b <= f) w[c++] = v[b++];
	while (i <= f) v[f--] = w[--c];
    free(w);
}

void merge_sort(Tipo v[], int i, int f) {
    if (i == f) return;
    int m = (i + f) / 2;
    merge_sort(v, i, m);
    merge_sort(v, m+1, f);
    intercala(v, i, m, f);
}

/*────────────────────────────────────────────────
 Quick Sort (determinístico/probabilístico)
────────────────────────────────────────────────*/
int particiona(Tipo v[], int i, int f) {
    //troca(v, i, aleatorio(i, f));
    int e = i+1, d = f;
    while (e <= d) {
        while (e <= d && v[e] <= v[i]) e++;
        while (v[d] > v[i]) d--;
        if (e < d) {
            troca(v, e, d);
            e++;
            d--;
        }
    }
    troca(v, i, d);
    return d;
}

void quick_sort(Tipo v[], int i, int f) {
    if (i >= f) return;
    int p = particiona(v, i, f);
    quick_sort(v, i, p-1);
    quick_sort(v, p+1, f);
}

/*────────────────────────────────────────────────
 Testes
────────────────────────────────────────────────*/
int main(void) {
    int n = 10000;
    Tipo *v = malloc(n * sizeof(Tipo));
    Tipo *w = malloc(n * sizeof(Tipo));
    
    for (int i = 0; i < n; i++) w[i] = aleatorio(0, n-1);
    
    for (int i = 0; i < n; i++) v[i] = w[i];
    printf("crescente: %s\n", crescente(v, n) ? "sim" : "nao");
    
    bubble_sort(v, n);
    printf("crescente: %s\n", crescente(v, n) ? "sim" : "nao");
    
	for (int i = 0; i < n; i++) v[i] = w[i];
    selection_sort(v, n);
    printf("crescente: %s\n", crescente(v, n) ? "sim" : "nao");
    
	for (int i = 0; i < n; i++) v[i] = w[i];
    insertion_sort(v, n);
    printf("crescente: %s\n", crescente(v, n) ? "sim" : "nao");
 
    for (int i = 0; i < n; i++) v[i] = w[i];
    merge_sort(v, 0, n-1);
    printf("crescente: %s\n", crescente(v, n) ? "sim" : "nao");
 
    for (int i = 0; i < n; i++) v[i] = w[i];
    quick_sort(v, 0, n-1);
    printf("crescente: %s\n", crescente(v, n) ? "sim" : "nao");
    
	return 0;
}
